homeassistant/components/light/__init__.py

```diff
@@ -228,6 +256,48 @@ async def async_handle_light_on_service(light, call):
         if ATTR_PROFILE not in params:
             profiles.apply_default(light.entity_id, params)
 
+        supported_color_modes = light.supported_color_modes
+        # Backwards compatibility: if an RGBWW color is specified, convert to RGB + W
+        # for legacy lights
+        if ATTR_RGBW_COLOR in params:
+            legacy_supported_color_modes = (
+                light._light_internal_supported_color_modes  # pylint: disable=protected-access
+            )
+            if (
+                COLOR_MODE_RGBW in legacy_supported_color_modes
+                and not supported_color_modes
+            ):
+                rgbw_color = params.pop(ATTR_RGBW_COLOR)
+                params[ATTR_RGB_COLOR] = rgbw_color[0:3]
+                params[ATTR_WHITE_VALUE] = rgbw_color[3]
+
+        # If a color is specified, convert to the color space supported by the light
+        # Backwards compatibility: Fall back to hs color if light.supported_color_modes
+        # is not implemented
+        if not supported_color_modes:
+            if (rgb_color := params.pop(ATTR_RGB_COLOR, None)) is not None:
+                params[ATTR_HS_COLOR] = color_util.color_RGB_to_hs(*rgb_color)
+            elif (xy_color := params.pop(ATTR_XY_COLOR, None)) is not None:
+                params[ATTR_HS_COLOR] = color_util.color_xy_to_hs(*xy_color)
+        elif ATTR_HS_COLOR in params and COLOR_MODE_HS not in supported_color_modes:
+            hs_color = params.pop(ATTR_HS_COLOR)
```

Comment:
You could go all wallrus:

```suggestion
        elif COLOR_MODE_HS not in supported_color_modes and (hs_color := params.pop(ATTR_HS_COLOR, None) is not None:
```

Reply:
It becomes 4 lines instead 2 after formatting though:
```py
elif (
            COLOR_MODE_HS not in supported_color_modes
            and (hs_color := params.pop(ATTR_HS_COLOR, None)) is not None
        ):
```